### PR TITLE
Keyboard input accessory ios

### PR DESCRIPTION
**Pull request by ekrivopaltsev**
Intuit uses React Native in presenting Tax interview process, one of the essential problems we found that not all IOS keyboard provides a place for "Next", "Done" buttons even when keyReturnType is set to Next or Done. As a result we could not traverse an interview form while keyboardType is "number" or "number" pad or should avoid few useful keyboard types. This is IOS issue not ReactNative, however in previous native implementation we were able to add inputAccessoryView to UITextField and put there Next/Done buttons. The proposed solution enable inputAccessoryView to be added to RCTTextField.

We debated two possible solution - one with exposing RCTText include files to the project and second without exposing, this solution does not require RCTTextField to be exposed to a ReactApplication and uses selectors instead if explicit RCTTextField methods.

**Test plan (required)**

Here is "number" keyboard appearance with requested returnKeyType = Done
image

Here how it looks after improvement, note that inputViewAccessory is formed by application and is not related to this commit.
image

Here is an application code which presents a test case.

### 1. Keyboard registration

```objc
 Class clz = NSClassFromString(@"RCTTextField");

  KeyboardAccessory keyboardHandler = ^UIView *(UITextField *textField) {
    ReactTextAdapter *kAdapter = [[ReactTextAdapter alloc] init:textField];
    return [kAdapter getInputAccessory];
  };
  

  SEL selector = NSSelectorFromString(@"registerForKeyboardType:handler:");
  if ([clz respondsToSelector:selector]) {
    [ clz performSelector:selector withObject:[NSNumber numberWithInt:UIKeyboardTypeNumberPad] withObject:keyboardHandler];
  }
```

### 2. Implementation for React TextAdapter  

```objc
@interface ReactTextAdapter :UIToolbar

-(instancetype) init:(UITextField *)textField;
-(UIView *)getInputAccessory;

@end

//
//  ReactTextAdapter.m
//  KeyboardTest
//
//  Created by Krivopaltsev, Eugene on 3/11/16.
//  Copyright © 2016 Facebook. All rights reserved.
//

#import "ReactTextAdapter.h"


@interface ReactTextAdapter()

@property (nonatomic) UITextField *textField;

//@property (nonatomic)UIToolbar *toolBar;
@property (nonatomic)UIBarButtonItem *flexible;

@end

@implementation ReactTextAdapter


-(instancetype) init:(UITextField *)textField {
  self = [super init];
  if (self) {
    self.textField = textField;
    
    [self setBarStyle:UIBarStyleDefault];
    self.translatesAutoresizingMaskIntoConstraints = NO;
    self.flexible = [[UIBarButtonItem alloc]
                     initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace
                     target:self action:nil];
   
    
    
    // since there is a next field, add a NEXT button to the picker
    if (self. textField. returnKeyType == UIReturnKeyNext) {
      NSString *nextText = @"Next";
      UIBarButtonItem *nextButton = [[UIBarButtonItem alloc] initWithTitle:nextText
                                                                   style:UIBarButtonItemStyleDone
                                                                  target:self
                                                                  action:@selector(next)];
      
    
      self.items = [[NSArray alloc] initWithObjects: self.flexible, nextButton, nil];
    }else if (self.textField. returnKeyType == UIReturnKeyContinue) {
    
      NSString *continueText = @"Continue";
      UIBarButtonItem *contButton = [[UIBarButtonItem alloc] initWithTitle:continueText
                                                                     style:UIBarButtonItemStyleDone
                                                                    target:self
                                                                    action:@selector(next)];
      
      self.items = [[NSArray alloc] initWithObjects: self.flexible, contButton, nil];
    }
    else if (self. textField.returnKeyType == UIReturnKeyDone) {
    
      NSString *doneText = @"Done";
      UIBarButtonItem *doneButton = [[UIBarButtonItem alloc] initWithTitle:doneText
                                                                     style:UIBarButtonItemStyleDone
                                                                    target:self
                                                                    action:@selector(done)];
      doneButton.enabled = YES;
      
      self.items = [[NSArray alloc] initWithObjects: self.flexible, doneButton, nil];
    }
    
  }
  return self;
}
-(void) done {
  [self passToReactNative];
  [self.textField resignFirstResponder];
}

-(void) next {
  [self passToReactNative];
}

-(void) passToReactNative {
  SEL selector = @selector(textFieldSubmitEditing);
  if ([self.textField respondsToSelector:selector]) {
    [self.textField performSelector:selector withObject:nil];
  }
}



-(UIView *)getInputAccessory {
  return self;
}
```

Test Project (without node_modules)
[KeyboardTest.zip](https://github.com/facebook/react-native/files/190206/KeyboardTest.zip)

### CODE REVIEW

**Comment by mkonicek** [comment]
@ekrivopaltsev Can you sing the CLA please?

**Comment by ekrivopaltsev** [comment]
Martin, 
I remember I signed, I tried to sign again and it seems FB site confirms that:
image

Please let me know if I miss something.

**Comment by mkonicek** [comment]
Ah OK, might be a glitch then (haven't seen that before).

@javache What do you think about this PR? (Nick might be too overwhelmed with PRs, trying to distribute the code review.)

File: Libraries/Text/RCTTextField.m [hunk 1]
```diff
@@ -279,4 +296,27 @@ - (BOOL)resignFirstResponder
...
+  
+  return textFieldAccessoryMapping;
+}
+
+/**
+ * The method to allow application register handle to produce inputAccessoryView for particular keyboard type
+ * handler presents instancdispatch_oncee of KeyboardAccessory type
+ */
+
++(void) registerForKeyboardType:(NSNumber *)type handler:(KeyboardAccessory)handler{
```
Comment (javache) [inline Libraries/Text/RCTTextField.m]: Declare this method in the header so you don't need to do the NSSelectorFromString hack.

The API here should use `UIKeyboardType`. Keep NSNumber as an implementation detail.

File: Libraries/Text/RCTTextField.m [hunk 2]
```diff
@@ -279,4 +296,27 @@ - (BOOL)resignFirstResponder
   return result;
 }
 
+
++(NSMutableDictionary*)getAcessoryMapping {
```
Comment (javache) [inline Libraries/Text/RCTTextField.m]: nit: call this just `accessoryMapping`. Objective-C assessors are not prefixed with get.

File: Libraries/Text/RCTTextField.m [hunk 3]
```diff
@@ -257,6 +267,13 @@ - (BOOL)canBecomeFirstResponder
 
 - (void)reactWillMakeFirstResponder
 {
+  
+  // This fragment was added to let application inject input accessory view if needed
+  //
+  KeyboardAccessory handler = [RCTTextField getAcessoryMapping][[NSNumber numberWithInt:self.keyboardType]];
```
Comment (javache) [inline Libraries/Text/RCTTextField.m]: Use @(self.keyboardType) instead of numberWithInt

**Comment by javache** [comment]
I'm not sure if this is really something we'd want to merge as we typically haven't tried too much to provide native API's to React elements (complexity increases really quickly with some configuration coming from native and some from JS). E.g.: think about how you would handle customizing the input accessory on a page basis beyond just the keyboardType.

**Comment by ekrivopaltsev** [comment]
Hi Peter, 

Couple comments on using selector instead of explicit method in .h file, I initially implemented it exactly by declaring new methods in .h file, but by default RCText library is not exposed to the application and that was the reason to use selectors instead , so only React package is exposed in the project include search path, I could definitely include explicit methods into RCTTextField.h, it is also simpler in such case to returnKey instead of wrapping into an object as performSelector with object requires an object and that was a way to wrap a primitive into an object.

Generally we had an issue for using Next/Done which is specific for particular types of keyboards and there is no such challenge on Android. If you have a simpler way to address the issue we faced please let me know. If a general stance to explicitly expose few new methods ( I think 2) within RCTTextField I could do it in few hours.
Thank you for your feedback.

**Comment by facebook-github-bot-8** [comment]
@ekrivopaltsev do you have any updates for this pull request? It's been a while so we wanted to check in and see if you've looked at the requested changes.

**Comment by mkonicek** [comment]
@ekrivopaltsev I'm trying to understand what needs to be done here but can't understand your comment very well. Are you saying it's no possible to concerns Pieter mentioned? In that case this PR should be closed.

**Comment by ekrivopaltsev** [comment]
Not at all, My point was by using selectors it was not required to add RCTText framework into a search path. My understanding FB  wants only to have React to be exposed. If it is not a limitation, I update my code.

**Comment by facebook-github-bot-2** [comment]
@ekrivopaltsev updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@nicklockwood would you mind taking a look at this pull request? It's been a while since the last commit was reviewed.

**Comment by ekrivopaltsev** [comment]
Is it possible to accelerate review process ? please let me know if any additional materials are required

**Comment by brentvatne** [comment]
Hi @ekrivopaltsev!

A few of us discussed this and I think this is not the direction we want to go with accessory views. The primary problem with this approach is that it requires you to write your accessory views using native code -- I think it's worth exploring how we can implement accessory views that are backed by React components, in JS. If you want to explore that approach it would be awesome! We really appreciate your effort in this PR though, thanks a bunch!

**Comment by brentvatne** [comment]
Apologies if the closed seemed aggressive -- feel free to continue to make your case here if you disagree with anything I said above.

**Comment by ekrivopaltsev** [comment]
Hi Brent,
I did not see your comments as aggressive, IMHO it is more Apple issue as some of the keyboard types DO NOT support return key, there is no such issue on Android.  This is  one of the cases where Android behaves better. The solution seems like transparent for the application and does not impact application React JS code. I am not sure if there is a simpler way to achieve the same behavior, please suggest one.  As I mentioned initially for "number" keyboards "Next" and "Done" do not appear on iPhone  and as a result an application either need to use suboptimal keyboard type or close and open keyboard for each form field. The solution I proposed (using accessory view) was used in the previous  application version (fully native).  Though I acknowledge that UITextField accessory view may open many additional capabilities.
Thanks

**Comment by brentvatne** [comment]
@ekrivopaltsev - I agree from a UX point of view this is something that you commonly want to do, but I also think that you will frequently want to customize what you can do in the accessory bar -- and with this implementation you'd need to write native code for that. I think what we want is to be able to implement the accessory view in JavaScript.

**Comment by henrikra** [comment]
Is `inputAccessoryView` coming to RN? I would love to have submit button when keyboardType is `phone-pad` :)

**Pull request closed without merging**